include/ipc/ipc_service.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_IPC_SERVICE_IPC_SERVICE_H_
+#define ZEPHYR_INCLUDE_IPC_SERVICE_IPC_SERVICE_H_
+
+#include <stdio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief IPC Service API
+ * @defgroup ipc_service_api IPC service APIs
+ * @{
+ */
+
+/** @brief Event callback structure.
+ *
+ *  It is registered during endpoint registration.
+ *  This structure is part of the endpoint configuration.
+ */
+struct ipc_service_cb {
+	/** @brief Bind was successful.
+	 *
+	 *  @param priv Private user data.
+	 */
+	void (*bound)(void *priv);
+
+	/** @brief New packet arrived.
+	 *
+	 *  @param data Pointer to data buffer.
+	 *  @param len Length of @a data.
+	 *  @param priv Private user data.
+	 */
+	void (*received)(const void *data, size_t len, void *priv);
+
+	/** @brief An error occurred.
+	 *
+	 *  @param message Error message.
+	 *  @param priv Private user data.
+	 */
+	void (*error)(const char *message, void *priv);
+};
+
+/** @brief Endpoint instance.
+ *
+ *  Content is not important for user of the API.
+ *  It is implemented in a specific backend.
+ */
+struct ipc_ept;
+
+/** @brief Endpoint configuration structure. */
+struct ipc_ept_cfg {
+
+	/** Name of the endpoint. */
+	const char *name;
+
+	/** Id of the target processor. */
+	int target_core_id;
```

Comment:
In relation to this comment https://github.com/zephyrproject-rtos/zephyr/pull/36120#discussion_r650008407 I'm a bit confused about this.

An endpoint is part of an RPMsg instance and an RPMsg instance can only be used for IPC between two cores. So what is exactly the reason to have this at all?

The problem is that to be able to put two cores in communication we need a shared memory region and each RPMsg instance is only using one shared memory region so it is clear that this API is indeed only able to put two cores in communication using multiple channel / endpoints.

So my follow-up question is: how do we enable real multi-core (>2) communication? Because AFAICT `CONFIG_RPMSG_MULTI_INSTANCE_SHM_BASE_ADDRESS` is actually restricting the range of possible cores to exactly 2.

Reply:
The `target_core_id` field in the IPC-Service API was added to work with backends that support communication between multiple cores. 

That is right, RPMsg is used for IPC between two cores. Currently the backend implementation using RPMsg does not support working with multiple cores. 

The `target_core_id` field was intended to specify between which cores we are creating the endpoint. That is, the current core and the target core. It is still core-to-core communication, but within the selected two cores. 

It seems to me that in order to implement something like this in the future, separate regions would have to be allocated of shared memory for each pair of cores. The memory would have to be properly described, e.g. in a *.dts file for each core. 

Since currently the `target_core_id` field is not used, it can be removed. In the future when there is a need to create communication channels between multiple cores the IPC Service API will need to be extended.

Reply:
> An endpoint is part of an RPMsg instance and an RPMsg instance can only be used for IPC between two cores. So what is exactly the reason to have this at all?

I think naming is confusing here. There are IPC Service endpoints, RPMsg instances, and RPMsg endpoints.

I understand it's like you described with RPMsg instances and RPMsg endpoints. There can be one or more RPMsg instances between two cores and each of those instances can include multiple RPMsg endpoints. To enable RPMsg between cores A and B you need separated instance than between cores B and C.

However, IPC Service has only one instance in a single core. This instance can include multiple IPC Service endpoints. Each of these endpoints can be implemented as a separated RPMsg instance (or some of them can share RPMsg instances, but use separated RPMsg endpoints). And this way, IPC Service endpoints from single instance can be used to communicate with different cores (one endpoint to core A, another to core B). To precisely address IPC Service endpoint, you need information about destination core, to select correct RPMsg instance for given IPC Service endpoint.

Reply:
@maje-emb 

> The `target_core_id` field in the IPC-Service API was added to work with backends that support communication between multiple cores.

Is there any backend able to support multi-core (1 to X)  communication using one single shared memory region?

> It seems to me that in order to implement something like this in the future, separate regions would have to be allocated of shared memory for each pair of cores. The memory would have to be properly described, e.g. in a *.dts file for each core.

More precisely each DTS file for each core should describe a shared memory region for each RPMsg instance (X shared memory regions for X core-to-core IPC).

TBH this is a pretty big limitation IMO. I don't have any particular objection to see this PR merged but I would prefer to see an issue opened already where we can discuss how to extend this service to make it able to support real multi-core IPC.

@hubertmis 

> I think naming is confusing here. There are IPC Service endpoints, RPMsg instances, and RPMsg endpoints.

Yes, sorry. I should have spent more time on the correct naming.

> To enable RPMsg between cores A and B you need separated instance than between cores B and C.

Ok, but AFAICT this is not possible with the current API because currently you can only have one single RPMsg instance per core (aka one single shared memory region), so the B core have to decide if it wants to communicate with A or C exclusively. Is this correct?

Reply:
> Ok, but AFAICT this is not possible with the current API because currently you can only have one single RPMsg instance per core (aka one single shared memory region), so the B core have to decide if it wants to communicate with A or C exclusively. Is this correct?

I hope this API does not limit backend implementation in such way. I envision potential IPC Service backend implementation which could create n RPMsg instances based on DTS entries for IPC (containing shared memory regions, target CPU, IPM instances) during system initialization. Then IPC Service user does not have to worry about these hardware details. User could simply select `target_cpu` and `endpoint_name`. It would be backend responsibility to map `(target_cpu, endpoint_name)` tuple to given RPMsg instance and endpoint.

What do you think? Could it work?

Reply:
yes, this is exactly my point. We should extend this API (maybe in a future PR) to get all the needed information from DTS and build all the different RPMsg instances from there. How to do this is up for discussion though, this is why I was suggesting to open an issue where we can discuss all the gory details about this.

Reply:
> Is there any backend able to support multi-core (1 to X) communication using one single shared memory region?

At this time, the backend does not support 1 to X communication. Currently there is one backed prepared that uses RPMsg. Others may be created in the future.

Reply:
> So I'll remove `target_core_id` field a the IPC-Service API. Because it's unnecessary.

I agree. We can re-introduce this field later after the proper discussion @carlocaione suggested.

Reply:
Fixed.